### PR TITLE
Add initial support for STM32C0 Series with Nucleo C031C6 board

### PR DESCRIPTION
**Pull request by benjaminbjornsson**
This is an attempt to add support for the new STM32C0 Series.

I've tested the functionality using the following samples:
- samples/basic/blinky
- samples/basic/button
- samples/hello_world

I've made this a draft since I'm not really happy with the solution in 'clock_stm32_ll_common.c' and could use some help with a better solution.

### CODE REVIEW

**Comment by benjaminbjornsson** [comment]
@erwango @ABOSTM @FRASTM What do you think, should there be a Kconfig or macro for the backup domain access? Or am I missing something.

**Comment by erwango** [comment]
Thanks @benjaminbjornsson for this work. We were just about to start :-).
We'll provide feedback during this week

**Comment by benjaminbjornsson** [comment]
> Thanks @benjaminbjornsson for this work. We were just about to start :-).
> We'll provide feedback during this week

Thanks @erwango, hope I’m not getting in your way with this PR.

File: west.yml [hunk 1]
```diff
@@ -129,7 +129,7 @@ manifest:
       groups:
         - hal
     - name: hal_stm32
-      revision: c36136b253a46ed8dd63525eeee659393047ce83
+      revision: 7021c43bf97d075808470b7f63da9e33ce7c01a4
```
Comment (erwango) [inline west.yml:132]: Please use the following for now.
```
revision: pull/161/head
```
Once HAL PR is merged, we'll use the definitive SHA1

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 2]
```diff
@@ -598,11 +598,13 @@ static void set_up_fixed_clock_sources(void)
 
 		z_stm32_hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);
 
+#if !defined(CONFIG_SOC_SERIES_STM32C0X)
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_common.c:601]: I suggest:
```
#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP)
```

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 3]
```diff
@@ -704,7 +708,11 @@ int stm32_clock_control_init(const struct device *dev)
 
 	/* If freq increases, set flash latency before any clock setting */
 	if (old_flash_freq < CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) {
+#if defined(CONFIG_SOC_SERIES_STM32C0X)
+		UTILS_SetFlashLatency(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+#else
 		LL_SetFlashLatency(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+#endif /* CONFIG_SOC_SERIES_STM32C0X */
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_common.c:715]: This is clearly a bug in LL API.
Please add a new commit to the HAL PR which add `LL_SetFlashLatency` as a public version of `UTILS_SetFlashLatency`.
When doing so, please add reference to the patch in the `stm32cube/stm32c0xx/README` using internal bug reference: 146704.

**Review by FRASTM — COMMENTED** [review]
LGTM

File: west.yml (hunk 1 above)
Comment (benjaminbjornsson) [reply west.yml:132]: Neat, thanks!

File: drivers/clock_control/clock_stm32_ll_common.c (hunk 3 above)
Comment (benjaminbjornsson) [reply drivers/clock_control/clock_stm32_ll_common.c:715]: Fixed

File: boards/arm/nucleo_c031c6/arduino_r3_connector.dtsi [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+			   <14 0 &gpioa 9 0>,	/* D8 */
+			   <15 0 &gpioc 7 0>,	/* D9 */
+			   <16 0 &gpiob 0 0>,	/* D10 */
+			   <17 0 &gpioa 7 0>,	/* D11 */
+			   <18 0 &gpioa 6 0>,	/* D12 */
+			   <19 0 &gpioa 5 0>,	/* D13 */
+			   <20 0 &gpiob 9 0>,	/* D14 */
+			   <21 0 &gpiob 8 0>;	/* D15 */
+	};
+};
```
[the remark this answers is not in the document]
Comment (benjaminbjornsson) [reply boards/arm/nucleo_c031c6/arduino_r3_connector.dtsi:36]: I've added arduino_serial, the other two will have to wait for when I2C and SPI is supported.

**Comment by Desvauxm-st** [comment]
I test on nucleo-c031c6 with success all the following test:

sample/hello_world  ok
sample/basics/blinky  ok
sample/basics/button  ok (irq call back ok).
tests/drivers/gpio/gpio_get_direction ok
tests/drivers/gpio/gpio_hogs ok
tests/drivers/hwinfo/api ok
tests/drivers/uart/uart_basic_api ok
tests/drivers/console ok
tests/drivers/clock_control/clock_control_api ok
tests/drivers/clock_control/ stm32_clock_configuration\stm32_common_devices ok

**Comment by erwango** [comment]
@benjaminbjornsson You can move PR to "Ready to review" I guess

File: drivers/clock_control/clock_stm32_ll_common.c (hunk 2 above)
Comment (benjaminbjornsson) [reply drivers/clock_control/clock_stm32_ll_common.c]: Fixed

**Review by FRASTM — DISMISSED** [review]

**Comment by Desvauxm-st** [comment]
@benjaminbjornsson 
Do you plan to add other
"basic" IPs like  :
DAC, SPI, I2C, IWDG, WWDG, ADC, Timers (PWM, counters), RTC ?

**Comment by benjaminbjornsson** [comment]
> @benjaminbjornsson
> Do you plan to add other
> "basic" IPs like :
> DAC, SPI, I2C, IWDG, WWDG, ADC, Timers (PWM, counters), RTC ?

Yes, that’s the plan. I thought it would be a good idea to keep this initial PR as small as possible and add the other things in follow up PRs. I have all three dev-kits and the plan has been to have them all supported.

**Review by FRASTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**